OGCAPI-Records for GN5

davidblasby opened the pull request:
I'm adding this early so people can see what I am working on.

- [x] Initial work getting the OGC's OpenAPI (.yaml) working
- [x] Initial test cases for the model
- [x] Implementation of all the services
- [x] Testing 

[Summary PPT](https://github.com/user-attachments/files/18443641/gn5-ogc-records.pdf)


Please see the included .YAMLs.  These are originally from the [ogc's ogcapi-records github](https://github.com/opengeospatial/ogcapi-records/tree/master/core/openapi) repo.

I've modified them so they will produce "good" java code. There are quite a few minor changes - see the `README.md` in some of the directories.  Most of the changes I made don't affect the API.  

However, a few do make "real" API changes - mostly on the output side.  For example;

* IDs are always Strings (in the spec they can be a number or string).  For parsing the .json, numbers will automatically be converted to strings
* `catalog#recordType` is always a `List<String>` (in the spec they can be a single string or string array)
* I've modified how the GeoJSON parsing happens.  This doesn't affect the model, however, you must ensure that you set the sub-class (i.e. `OgcApiRecordsPolygonDto.java`) `type` (it not automatically done).

More details are in the `README.md` and comments at the top of the `.yaml` files.

I've added test cases for the seralizing/deserializing:
* examples from the OGCAPI spec
* examples that test some of the changes I made to the model

david-blasby committed:
initial work on converting ogcapi's openapi .yaml to usable java using openapi-generator

davidblasby commented:
@josegar74 @fxprunayre - please take a look. 

I will be looking at doing some actual implementations next.

david-blasby committed:
added content negotiation, better json output, trivial HTML output

david-blasby committed:
format changes

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/geonetwork/geonetwork?pullRequest=74) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/geonetwork/geonetwork?pullRequest=74) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/geonetwork/geonetwork?pullRequest=74) it.</sub>

david-blasby committed:
initial '/catalogs/{catalogid}' implementation'

david-blasby committed:
mostly working

david-blasby committed:
added conformance and links

david-blasby committed:
bug fixes, new features, fixed test cases, reformated code

david-blasby committed:
undo some format changes

david-blasby committed:
merge origin/main

david-blasby committed:
from origin/main - merge conflict resolution

david-blasby committed:
reformat all the pom.xml

david-blasby committed:
initial documenation for ogcapi-records

david-blasby committed:
remove old ogcapi-records

davidblasby commented:
@fxprunayre @josegar74 @juanluisrp - This is getting to the point of being usable now.
I just added documentation on how to setup a multi-portal (multi-collection) with "extra" collection metadata.

I havent done much testing, but it appears to be working.  

Please review.  I've asked @juanluisrp to make a deployment.

david-blasby committed:
add some more test cases

david-blasby committed:
formating changes

davidblasby commented:
Added PDF/PPT summary in the description that goes through a lot of technical details - also, good info about GN5.

david-blasby committed:
changes for working with QGIS

david-blasby committed:
changes for working with various clients

david-blasby committed:
for -Dqa

davidblasby commented:
Ok, I've tested this against a few clients.  Each of the clients has a few "quirks" - I updated GN5 OGCAPI-Records to account for this.

1. GeoCat's OGCAPI-Records (originally for geonetwork-microservices)
2. QGIS - MetaSearch plugin.  Use a specific collection as the url (i.e. http://localhost:7979/ogcapi-records/collections/subportal)
3. STAC Browser -  https://radiantearth.github.io/stac-browser 
4. Camp2Camp Info - https://camptocamp.github.io/ogc-client/#/demo (you need to host GN5 with HTTPS or use something like http://ngrok.com)
5. Open Geo Groep -  https://opengeogroep.github.io/ogc-api-features-testclient/src/index.html

david-blasby committed:
better application.yml

david-blasby committed:
added minor comments

david-blasby committed:
add query string -> queryable

davidblasby commented:
I noticed that all the clients (except my Angular) do not handle queryables.

I've added support for queryables in the `q` (query string).

http://localhost:7979/ogcapi-records/collections/subportal/items?q=contacts:geocat%20portal

This will search for "portal" with the `contacts=geocat` queryable.

This way you can use the queryables anywhere.

david-blasby committed:
add the elastic IndexRecord to output (records and catalogs).  also, its in the openapi swagger document

davidblasby commented:
added the elastic IndexRecord to catalog (i.e. `landing page` and `/collections` and `/collections/<collectionid>` and the individual records). 

I've also added this into the OpenAPI Swagger document, and it should keep itself updated when `IndexRecord.java` is modified (will need to build `src/module/ogcapi-records`).

david-blasby committed:
update documentation

davidblasby commented:
updated documentation

david-blasby committed:
reorg of docs + added technical documentation

davidblasby commented:
@josegar74 @fxprunayre - will merge this monday.

fxprunayre commented on docs/manual/docs/OGCAPI-Records/develop.md:
```diff
@@ -0,0 +1,110 @@
+
+## History
+
+The [GeoServer Microservices Project](https://github.com/geonetwork/geonetwork-microservices) had an OGCAPI-Records implementation.  This was used as a basis for the GN5 implementation.
```
```suggestion
The [GeoNetwork Microservices Project](https://github.com/geonetwork/geonetwork-microservices) had an OGCAPI-Records implementation.  This was used as a basis for the GN5 implementation.
```

davidblasby committed:
Update docs/manual/docs/OGCAPI-Records/develop.md

Co-authored-by: François Prunayre <fx.prunayre@gmail.com>

davidblasby commented:
@fxprunayre - thanks for the geoserver catch!  Oops!

davidblasby commented:
@fxprunayre @josegar74  - I'd like to merge this ASAP.  Its bit rotting and making other work more difficult.

fxprunayre committed:
Update OgcApiGeoJsonConverter.java

Tentative fix for 

```
{"code":"500","description":"Cannot invoke \"java.time.OffsetDateTime.toString()\" because \"x\" is null"}
```

fxprunayre commented:
Tested on a quite large catalog and works @davidblasby but I noticed some issues:
* on items
```
{"code":"500","description":"Cannot invoke \"java.time.OffsetDateTime.toString()\" because \"x\" is null"}
```
but with no details in the log so hard to track where this comes from. You can double check https://github.com/davidblasby/geonetwork/commit/6ce65d712fba180767c6e8b4db452bfc91a9d3ff


* We may have to support UUID with special characters

```
{"code":"500","description":"java.net.URISyntaxException: Illegal character in path at index 113: http:/localhost:7979/ogcapi-records/collections/b08fe709-1ced-4a07-8edf-06aa6ccdf2e3/items/SDN_CPRD_1568_KN172L06 SeaBeam 3012-P1 Bathymetry Grid?f=geojson"}
```

* Requesting large response, I'm getting

```
{"code":"500","description":"node: http://localhost:9200/, status: 200, [es/search] Failed to decode response"}
```

* in http://localhost:7979/doc/api/swagger-ui/index.html the server URL is wrong

![image](https://github.com/user-attachments/assets/8f3081ff-eeef-404f-8a9c-2f84e7ce1121)



Is there any way to get exception details?

david-blasby commented:
Hi, Francois,

There is code to handle the exception in `src/modules/ogcapi-records/ogcapi-records-implementation/src/main/java/org/geonetwork/ogcapi/controllerexceptions/GlobalExceptionHandler.java`.  We should probably have it log the error (and stacktrace).

The special chars (like ' ') in the UUID probably just needs to make sure that we URL-encode the links.  How are you getting to that page?  If you are copy-and-pasting it to the browser location bar, then we probably need to make sure that spring boot will automatically fix that url before internal dispatch. 
However, if you are getting that link in the JSON, then we should just make sure that we are URL-encoding the links.

I will look at this tomorrow:

1. log exceptions
2. URL-Encode links
3. test and see what happens when you have `space` chars in the browser URL-bar.
4. look at the swagger UI output.  I expect this has something to do with how swagger is building the UI based on the various endpoints.  When I looked at the json output, i didn't see these problems...

Thanks for your feedback,
Dave

david-blasby committed:
fix url encoding of 'bad' uuid strings, log controller error messages

david-blasby commented:
pushed commit for;

a) turning on controller error logging
b) better url encoding of uuid/collectionid
c) problem was in the link generation - not in the controller endpoints

 @fxprunayre  - can you give me some context for the "Requesting large response, I'm getting".  What are you doing?  Can you give me the exception stack trace (and caused by).  You should be able to get it easily with the latest commit.

I will look at the swagger stuff tomorrow.

david-blasby committed:
swagger ui ('try it out') so it doesn't inject the host twice

davidblasby commented:
The swagger stuff was a quick fix - not sure why it was a problem...

@fxprunayre - can you give me some context for the "Requesting large response, I'm getting". What are you doing to cause it? Can you give me the exception stack trace (and caused by)? You should be able to get it fairly easily with yesterday's erorr-logging commit.

davidblasby committed:
Merge branch 'main' into _openapi

david-blasby committed:
fix issues with changes to  branch

davidblasby commented:
@fxprunayre  - I did the merge commit inside github because it "looked easy."  Turns out it was more complex. 

FIXED.

davidblasby commented:
@fxprunayre @josegar74 - I would like to merge this.

PascalLike commented on src/modules/ogcapi-records/ogcapi-records-implementation/src/main/java/org/geonetwork/ogcapi/configuration/WebConfig.java:
```diff
@@ -0,0 +1,74 @@
+ * <p>This contains some beans and some overrides for setting up spring.
+ */
+@Configuration
+public class WebConfig implements WebMvcConfigurer {
```
Did you consider the option of using the annotation @EnableWebMvc?
 https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/servlet/config/annotation/WebMvcConfigurer.html

davidblasby replied:
yes - this was the recommended way to setup the ContentNegotiationConfigurer and provide an objectmapper.

Is there a better way?

davidblasby replied:
I am going to merge this - but its actively being worked on.  If you have a better way, I can change it.

fxprunayre commented on src/shared/ogcapi-records/src/main/java/org/geonetwork/ogcapi/service/validation/SimpleValidators.java:
```diff
@@ -0,0 +1,29 @@
+        if (isBlank(catalogId)) {
+            throw new Exception("no catalogId!");
+        }
+        if (!catalogId.matches("^[a-zA-Z0-9_-]+$")) {
```
This is a bit more strict than what GN4 does. We may end-up with message like

```
{"code":"500","description":"java.lang.Exception: catalogId 'metawal.wallonie.be' is invalid!"}
```
when the main catalogue id is set in the settings:

<img width="524" height="195" alt="Image" src="https://github.com/user-attachments/assets/531474eb-01b4-47a7-a8a1-2aafbcc49e0a" />



For portals, we have a client side check on portal name

<img width="524" height="195" alt="Image" src="https://github.com/user-attachments/assets/47ff38df-9370-437c-b530-5cf1890cc63e" />

but none on API or DB side.

Should we be more strict in all places?

PascalLike replied:
These restrictions are not from the OGC API specification directly, but from the fact that collectionId (as well as itemId, which are metadata identifiers) are used in URLs. Generally, it would be sufficient to use characters that can be used in URLs... Then I think the period `.` could be added to the regex. What do you think @davidblasby ?

fxprunayre replied:
It makes me think of this old discussion in GN4 also https://github.com/geonetwork/core-geonetwork/pull/5736

In OGC API spec `collectionId` is `string` or there are some constraints?

davidblasby replied:
I found that accepting anything as an ID would be problematic from a security perspective.

Feel free to open this up to a few more characters (like ".").  

Records can have some craziness for their IDs.

Here is what google-ai recommends:

```
Best Practices for "Safe" IDs

To ensure your API works across all clients and servers, follow these "unwritten" rules:

- Stick to URL-Safe Characters: Use a-z, 0-9, -, and _. Avoid spaces and special symbols.

- [Kebab-case or Snake_case: These are standard for {collectionId} (e.g., natural-parks or water_wells).

- UUIDs for Features: If your data doesn't have natural IDs, UUIDs (e.g., 550e8400-e2...) are a safe, standard way to ensure uniqueness across systems.

- Avoid Lengthy IDs: Extremely long IDs can sometimes hit URL character limits in older browsers or proxy servers. 
```
